File: extractor/src/Item/DataStructures/Item.ts
import { Expose, Type } from "class-transformer";
import { ArrayEqual } from "../../CompareUtils/ArrayEqual.js";
import { RecordObject } from "../../Database/RecordObject.js";
import { ItemMoveInfo } from "./ItemMoveInfo.js";
import { SqlField } from "../../SqlConverter/Decorators/SqlField.js";
import { SqlNestedField } from "../../SqlConverter/Decorators/SqlNestedField.js";

/**
 * Represents a Item in the tool.
 */
export class Item implements RecordObject {
	/**
	 * The File Version that originated this object
	 */
	@Expose()
	@SqlField()
	public _FileVersion: number = -1;

	/**
	 * Item ID
	 */
	@Expose()
	@SqlField()
	public Id: number = 0;

	@Expose()
	@SqlField()
	public IdentifiedName: string = "<<Incomplete Item>>";

	@Expose()
	@SqlField({
		transform: (value: string[]) => value.join('\n'),
		outType: () => String,
	})
	public IdentifiedDescription: string[] = [];

	@Expose()
	@SqlField()
	public IdentifiedSprite: string = "";

	@Expose()
	@SqlField()
	public UnidentifiedName: string = "";

	@Expose()
	@SqlField({
		transform: (value: string[]) => value.join('\n'),
		outType: () => String,
	})
	public UnidentifiedDescription: string[] = [];

	@Expose()
	@SqlField()
	public UnidentifiedSprite: string = "";

	@Expose()
	@SqlField()
	public SlotCount: number = 0;

	@Expose()
	@SqlField()
	public IsBook: boolean = false;

	@Expose()
	@SqlField()
	public CanUseBuyingStore: boolean = false;

	@Expose()
	@SqlField()
	public CardPrefix: string = "";

	@Expose()
	@SqlField()
	public CardPostfix: string = "";

	@Expose()
	@SqlField()
	public CardIllustration: string = "";

	@Expose()
	@SqlField()
	public ClassNum: number = 0;

	@Expose()
	@Type(() => ItemMoveInfo)
	@SqlNestedField()
	public MoveInfo: ItemMoveInfo = new ItemMoveInfo();

	public getId(): string {
		return this.Id.toString();
	}

	public getFileVersion(): number {
		return this._FileVersion;
	}

	public equals(other: Item): boolean {
		// _FileVersion is not checked, if 2 versions exists but the record is the same, we don't care.
		return (
			this.IdentifiedName === other.IdentifiedName
			&& ArrayEqual.isEqual(this.IdentifiedDescription, other.IdentifiedDescription)
			&& this.IdentifiedSprite === other.IdentifiedSprite
			&& this.UnidentifiedName === other.UnidentifiedName
			&& ArrayEqual.isEqual(this.UnidentifiedDescription, other.UnidentifiedDescription)
			&& this.UnidentifiedSprite === other.UnidentifiedSprite
			&& this.SlotCount === other.SlotCount
			&& this.IsBook === other.IsBook
			&& this.CanUseBuyingStore === other.CanUseBuyingStore
			&& this.CardPrefix === other.CardPrefix
			&& this.CardPostfix === other.CardPostfix
			&& this.CardIllustration === other.CardIllustration
			&& this.ClassNum === other.ClassNum
			&& this.MoveInfo.equals(other.MoveInfo)
		);
	}
}

File: extractor-go/internal/domain/Item.go
package domain

import (
	"database/sql"
)

type ItemMoveInfo struct {
	CanDrop               bool
	CanTrade              bool
	CanMoveToStorage      bool
	CanMoveToCart         bool
	CanSellToNpc          bool
	CanMail               bool
	CanAuction            bool
	CanMoveToGuildStorage bool
	CommentName           string
}

func NewItemMoveInfo() ItemMoveInfo {
	return ItemMoveInfo{
		CanDrop:               true,
		CanTrade:              true,
		CanMoveToStorage:      true,
		CanMoveToCart:         true,
		CanSellToNpc:          true,
		CanMail:               true,
		CanAuction:            true,
		CanMoveToGuildStorage: true,
	}
}

func (i *ItemMoveInfo) Equals(other ItemMoveInfo) bool {
	return (i.CanDrop == other.CanDrop &&
		i.CanTrade == other.CanTrade &&
		i.CanMoveToStorage == other.CanMoveToStorage &&
		i.CanMoveToCart == other.CanMoveToCart &&
		i.CanSellToNpc == other.CanSellToNpc &&
		i.CanMail == other.CanMail &&
		i.CanAuction == other.CanAuction &&
		i.CanMoveToGuildStorage == other.CanMoveToGuildStorage &&
		i.CommentName == other.CommentName)
}

type Item struct {
	PreviousHistoryID       sql.NullInt32
	HistoryID               sql.NullInt32
	ItemID                  int32
	FileVersion             int32
	IdentifiedName          sql.NullString
	IdentifiedDescription   sql.NullString
	IdentifiedSprite        sql.NullString
	UnidentifiedName        sql.NullString
	UnidentifiedDescription sql.NullString
	UnidentifiedSprite      sql.NullString
	SlotCount               int8
	IsBook                  bool
	CanUseBuyingStore       bool
	CardPrefix              sql.NullString
	CardIsPostfix           bool
	CardIllustration        sql.NullString
	ClassNum                sql.NullInt32
	MoveInfo                ItemMoveInfo
	Deleted                 bool
}

func NewItem(itemID int32, fileVersion int32) Item {
	return Item{
		ItemID:      itemID,
		FileVersion: fileVersion,
		MoveInfo:    NewItemMoveInfo(),
	}
}

func (i *Item) Equals(otherItem Item) bool {
	// FileVersion is not checked, if the file has changed but the Item is the same, we don't care.
	return (i.ItemID == otherItem.ItemID &&
		i.IdentifiedName == otherItem.IdentifiedName &&
		i.IdentifiedDescription == otherItem.IdentifiedDescription &&
		i.IdentifiedSprite == otherItem.IdentifiedSprite &&
		i.UnidentifiedName == otherItem.UnidentifiedName &&
		i.UnidentifiedDescription == otherItem.UnidentifiedDescription &&
		i.UnidentifiedSprite == otherItem.UnidentifiedSprite &&
		i.SlotCount == otherItem.SlotCount &&
		i.IsBook == otherItem.IsBook &&
		i.CanUseBuyingStore == otherItem.CanUseBuyingStore &&
		i.CardPrefix == otherItem.CardPrefix &&
		i.CardIsPostfix == otherItem.CardIsPostfix &&
		i.CardIllustration == otherItem.CardIllustration &&
		i.ClassNum == otherItem.ClassNum &&
		i.MoveInfo.Equals(otherItem.MoveInfo))
}
